kernel/arch/riscv/paging.c

```diff
@@ -264,40 +951,268 @@ void *init_early_mapping(ulong fdt_paddr)
    return (void *)(kernel_pa - MB);
 }
 
+static inline bool
+in_big_page(pdir_t *pdir, void *vaddrp)
+{
+   page_table_t *pmd;
+   const ulong vaddr = (ulong) vaddrp;
+   page_t *e;
+
+   for (int level = RV_PAGE_LEVEL; level > 1; level--) {
+
+      e = &pdir->entries[PTE_INDEX(level, vaddr)];
+
+      if (!e->present)
+         return false;
+
+      /* 1 GB super big page, not we want*/
+      if (e->raw & _PAGE_LEAF)
+         return false;
+
+      pmd = PA_TO_LIN_VA(e->pfn << PAGE_SHIFT);
+      pdir = pmd;
+   }
+
+   e = &pdir->entries[PTE_INDEX(1, vaddr)];
+
+   return (e->present) && (e->raw & _PAGE_LEAF);
+}
+
+static void
+set_big_page_attr(pdir_t *pdir,
+                  void *vaddrp,
+                  ulong attr)
+{
+   page_table_t *pmd;
+   const ulong vaddr = (ulong) vaddrp;
+   ulong *pte;
+   page_t *e;
+
+   for (int level = RV_PAGE_LEVEL; level > 1; level--) {
+
+      e = &pdir->entries[PTE_INDEX(level, vaddr)];
+
+      ASSERT((e->present) && !(e->raw & _PAGE_LEAF));
```

Comment:
I guess here as well a macro to check for leaf page alignment would help..

Reply:
fixed!